### PR TITLE
feat: add `SmallestDenomination` MoneyOption

### PR DESCRIPTION
**Pull request by david-zw-liu**
## Changes
1. add a new rounding mode, `ROUND_HALF_UP`
2. add `smallestDenomination` option

### CODE REVIEW

**Review by kken339039 — APPROVED** [review]

File: money.go [hunk 1]
```diff
@@ -105,14 +121,32 @@ func alignRoundingMode(m *Money, ma []*Money) MoneyOption {
...
+	smallestDenomination := 1.0
+	if m.smallestDenomination != 0 {
+		smallestDenomination = float64(m.smallestDenomination)
+	}
+	value = value / smallestDenomination
+	return roundCentsWithExplicitMode(value, m.roundingMode) * smallestDenomination
 }
 
 // DEPRECATED: Use Round instead. There is no option not to round by mode
 func (m *Money) RoundByMode(value float64) float64 {
```
Comment (mgleon08) [inline money.go]: 這個好像 DEPRECATED 了，還需要留嗎? 需不需要順便都改掉?
cc @Seitk

File: money.go [hunk 2]
```diff
@@ -61,13 +69,14 @@ func NewFromAmount(dollars float64, isoCode string, options ...MoneyOption) *Mon
...
-		Label:          nm.Display(),
-		roundingMode:   RoundBankers, // Default Round Mode will be RoundBankers
+		money:                nm,
+		Cents:                nm.Amount(),
+		Dollars:              nm.AsMajorUnits(),
+		CurrencyIso:          nm.Currency().Code,
+		CurrencySymbol:       nm.Currency().Grapheme,
+		Label:                nm.Display(),
+		roundingMode:         RoundBankers, // Default Round Mode will be RoundBankers
+		smallestDenomination: 1,
```
Comment (Seitk) [inline money.go]: I think by default the smallestDenomination should be based on currency, maybe add that to the currencies config?

File: money.go [hunk 3]
```diff
@@ -105,14 +122,32 @@ func alignRoundingMode(m *Money, ma []*Money) MoneyOption {
...
+	if m.smallestDenomination != 0 {
+		return WithSmallestDenomination(m.smallestDenomination)
+	}
+	fm, isFound := lo.Find(ma, func(money *Money) bool {
+		return money.smallestDenomination != 0
+	})
+	if isFound {
+		return WithSmallestDenomination(fm.smallestDenomination)
+	}
+	return WithSmallestDenomination(1)
```
Comment (Seitk) [inline money.go]: same here, don't always assume 1

File: money.go [hunk 4]
```diff
@@ -105,14 +122,32 @@ func alignRoundingMode(m *Money, ma []*Money) MoneyOption {
 	return WithRoundingMode(RoundBankers)
 }
 
+func alignSmallestDenomination(m *Money, ma []*Money) MoneyOption {
+	if m.smallestDenomination != 0 {
+		return WithSmallestDenomination(m.smallestDenomination)
+	}
+	fm, isFound := lo.Find(ma, func(money *Money) bool {
+		return money.smallestDenomination != 0
```
Comment (Seitk) [inline money.go:133]: I think it's possibly to have smallest denomination as 0 for some currencies

File: money.go (hunk 1 above)
Comment (Seitk) [reply money.go]: yes but please check if we removed all the callers

File: money.go (hunk 4 above)
Comment (david-zw-liu) [reply money.go:133]: the smallest denomination should never be zero because it's the smallest unit of a currency.

File: money.go [hunk 5]
```diff
@@ -105,14 +124,32 @@ func alignRoundingMode(m *Money, ma []*Money) MoneyOption {
...
+	if isFound {
+		return WithSmallestDenomination(fm.smallestDenomination)
+	}
+	return WithSmallestDenomination(m.GetCurrency().smallestDenomination)
+}
+
 // Round money with rounding mode set
 func (m *Money) Round(value float64) float64 {
-	return roundCentsWithExplicitMode(value, m.roundingMode)
+	smallestDenomination := float64(m.GetCurrency().smallestDenomination)
```
Comment (Seitk) [inline money.go]: change `m.GetCurrency().smallestDenomination` to `m.smallestDenomination`

**Review by Seitk — APPROVED** [review]

**Pull request merged**